_posts/2021-03-21-clean-unit-tests-with-mockito.md

```diff
@@ -0,0 +1,375 @@
+what decreases your code's maintainability. It will save you much time debugging and doesn't let your team members guess
+what the intent of the test case is.
+
+* Do not concentrate all the Mockito declarations in the `setUp` method! Even tough the test cases are reduced to a
+  minimum, the readability suffers a lot. Besides, you must highly pay attention to not break any other test cases. Like
+  so, you avoid interfering other tests by overriding the setup.
+* Do not recycle your mocks! It's better to create new mocks for each test case. Otherwise, you might experience
+  unexpected behavior through interfering declarations.
+* Do not expect test cases to be always executed in the same order! Write everything that belongs to your test case into
+  the test method, so that a test case can be executed alone in your IDE or all together within you CI!
+* Do not mock Collections nor value types. It is better to use real instances and values!
+* Do not call `reset()` in a test procedure. This is considered as code smell.
+
+## Mockito Specialities and Restrictions
+
+In this section we want to point out important things which are nice to know.
+
+* *What types can I mock?* Mockito allows you to mock not only interfaces but also concrete classes.
+* *What is returned, if I don't declare a mock's behaviour?* Mockito returns `null` for reference objects, and the
+  default values for primitive data types (for example `0` for `int` and `false` for `boolean`)
+* *How many times Mockito returns a previously declared value?* Mockito returns always the same value, regardless of how
+  many times a method is called.
+* *Can I mock `final` classes?* No, final classes **can't** be mocked and neither final methods are mockable. This has
+  to do with the internal mechanism of how Mocktio creates the mock and the Java Language Specification.
+
+### Testing Error Handling with Mockito
+
+```java
+Mockito.when(cityRepository.find(expected.getId())).thenThrow(RuntimeException.class);
+```
+
+Mockito comes with a built-in mechanism to test your error handling. Instead of declaring a return value, advise Mockito
+to throw the expected exception. In case you throw checked exceptions, the compiler doesn't let you to throw checked
+exceptions, that are not declared on the method.
+
+### Mocktio and `void` Methods
+
+```java
+// Causes a compiler error
+Mockito.when(cityRepository.delete()).thenThrow(RuntimeException.class);
+```
+
+In case you want to declare a special behaviour for `void` methods, you must change your approach. The compiler doesn't
+like void methods in brackets, since they have no arguments. Change your approach to following:
+
+```java
+Mockito.doThrow(RuntimeException.class).when(cityRepository).delete(expected);
+```
+
+### Verify Method Calls
+
+```java
+Mockito.verify(cityRepository,Mockito.times(1)).delete(expected);
+```
+
+You can verify how many times a mock was called by simply use the built-in `verify` method. If the condition is not met,
```

Comment:
There's a lot of "you" in this article. I have replaced it with "we" in a couple places, but not all of them. Please replace you with "we" in the rest of the places.